contracts/dao/neutron-flashloans/src/contract.rs

```diff
@@ -0,0 +1,403 @@
+        msgs: vec![Any {
+            type_url: BANK_MSG_SEND_TYPE_URL.to_string(),
+            value: bank_send_msg.encode_to_vec(),
+        }],
+    };
+
+    // Then we wrap the authz message in a stargate message, because there is
+    // no custom support for authz messages in CosmWasm.
+    let stargate_authz_msg_exec: CosmosMsg<NeutronMsg> = CosmosMsg::Stargate {
+        type_url: AUTHZ_MSG_EXEC_TYPE_URL.to_string(),
+        value: Binary(authz_msg_exec.encode_to_vec()),
+    };
+
+    stargate_authz_msg_exec
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    match msg.id {
+        AUTHZ_BANK_SEND_REPLY_ID => {
+            // If we are here, the money is already on the borrower's account. This means that
+            // we can proceed to call the borrower's handler.
+
+            let config: Config = CONFIG.load(deps.storage)?;
+            let active_loan = must_get_active_loan(&deps)?;
+
+            let msg = CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: active_loan.borrower.to_string(),
+                msg: to_json_binary(&BorrowerInterface::ProcessLoan {
+                    // We sent the return address to the source address
+                    return_address: config.source,
+                    loan_amount: active_loan.amount,
+                    fee: active_loan.fee,
+                })
+                .unwrap(),
+                funds: vec![],
+            });
+            // We use the reply_on_success strategy (we want the transaction to be simply reverted
+            // in case of an error).
+            Ok(Response::new()
+                .add_submessage(SubMsg::reply_on_success(msg, BORROWER_HANDLER_REPLY_ID))
+                .add_attribute("action", "reply_authz_bank_send"))
+        }
+        BORROWER_HANDLER_REPLY_ID => {
+            // If we are here, the borrower smart contract has successfully executed the
+            // ProcessLoan message, and probably returned the funds to our contract together with
+            // the additional fee.
+
+            let config: Config = CONFIG.load(deps.storage)?;
+            let active_loan = must_get_active_loan(&deps)?;
+
+            // Check that the borrower returned the loan and paid the fee
+            check_expected_balances(deps.as_ref(), config.source, active_loan.expected_balances)?;
+
+            // Set the active loan to None, thus making ourselves ready for the next loan
+            ACTIVE_LOAN.save(deps.storage, &None)?;
+
+            Ok(Response::new().add_attribute("action", "reply_borrower_handler"))
+        }
+        _ => Err(ContractError::UnknownReplyID { id: msg.id }),
+    }
+}
+
+fn check_expected_balances(
+    deps: Deps,
+    source: Addr,
+    expected_balances: Vec<Coin>,
+) -> Result<(), ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
```

Comment:
The same about a potentially _infinite_ loop

Reply:
Acknowledged, will fix!